### PR TITLE
[Elemental] Echoes Of The Great Sundering

### PR DESCRIPTION
**Pull request by HawkCorrigan**
Should look a bit better this time around.
Refactored a bit and fixed some edge cases.

The logs I checked with looked good to me.

### CODE REVIEW

**Comment by HawkCorrigan** [comment]
image

File: src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js [hunk 1]
```diff
@@ -6,66 +6,118 @@ import ITEMS from 'common/ITEMS';
...
-        }
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    if (spellId !== SPELLS.EARTHQUAKE.id)
+      return;
+
+    if (this.combatants.selected.hasBuff(SPELLS.ECHOES_OF_THE_GREAT_SUNDERING_BUFF.id, event.timestamp)) {
+      this.buffedCastCounter++;
+      this.endtime = event.timestamp + ECHOES.PROC_DURATION + ECHOES.LAG_TOLERANCE;
+      this.state = 1;
```
Comment (MartijnHols) [inline src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js]: Please make constants at the top of your file for magic numbers to make it easier to find what they mean.

File: src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js [hunk 2]
```diff
@@ -6,66 +6,118 @@ import ITEMS from 'common/ITEMS';
...
     }
+    else {
+      if (event.timestamp < this.endtime) {
+        this.state = 2;
```
Comment (MartijnHols) [inline src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js]: Please make constants at the top of your file for magic numbers to make it easier to find what they mean.

File: src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js [hunk 3]
```diff
@@ -6,66 +6,118 @@ import ITEMS from 'common/ITEMS';
...
+      }
+    }
+  }
+
+  on_byPlayer_damage(event) {
+    if (event.timestamp > this.endtime)
```
Comment (MartijnHols) [inline src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js]: Please use curly brackets around if-statements.

File: src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js [hunk 4]
```diff
@@ -6,66 +6,118 @@ import ITEMS from 'common/ITEMS';
...
+      this.state = 0;
```
Comment (MartijnHols) [inline src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js]: Please make constants at the top of your file for magic numbers to make it easier to find what they mean.

File: src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js [hunk 5]
```diff
@@ -6,66 +6,118 @@ import ITEMS from 'common/ITEMS';
...
+        this.buffedTickCounter++;
+        break;
+      case 2:
+        if (this.buffedBaseDamageSum !== 0 && this.buffedTickCounter > 0) {
+          const averageBuffedBaseDamage = this.buffedBaseDamageSum / this.buffedTickCounter;
+          if (Math.abs(baseDamage / averageBuffedBaseDamage) < ECHOES.TRESHOLD_PERCENTAGE) {
+            this.buffedEarthquakeDamage += event.amount;
+            this.buffedBaseDamageSum += baseDamage;
+            this.buffedTickCounter++;
+            console.log(this.buffedBaseDamageSum);
```
Comment (MartijnHols) [inline src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js]: Please do not leave `console.log` statements active in production code. A common practice is to add a `debug` constant at the top of the file that toggles them on or off.

File: src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js [hunk 6]
```diff
@@ -6,66 +6,118 @@ import ITEMS from 'common/ITEMS';
...
 
-    on_byPlayer_applybuff(event) {
-        const spellId = event.ability.guid;
-        if (spellId === SPELLS.ECHOES_OF_THE_GREAT_SUNDERING_BUFF.id) {
-            this.echoesProcsCounter += 1;
+    const critMultiplier = (event.hitType === HIT_TYPES.CRIT) ? CRIT_MULTIPLIER : 1;
+    const baseDamage = event.amount / critMultiplier;
+
+    switch (this.state) {
+      case 0:
```
Comment (MartijnHols) [inline src/Parser/Shaman/Elemental/Modules/Items/EchoesOfTheGreatSundering.js]: Please make constants at the top of your file for magic numbers to make it easier to find what they mean.

**Pull request merged**